### DISCUSSION TITLE
Hydrogen on Windows and MIDI

### DISCUSSION THREAD

**Discussion by rex007can** [hydrogen-music/hydrogen#1025] [Hydrogen on Windows]
Hello.
I've been using Hydrogen for years now but always on Linux.
I just started trying to use it on Windows and I'm having problems setting it up for MIDI.
I have all my stuff set up (pocket midi, loopMIDI, Cakewalk etc.)

I want Hydrogen to accept generic Start/Stop/resume/pause MIDI commands, but it just isn't receiving anything other than CC commands.

When I use PocketMIDI to generate MIDI events like start/stop and have Hydrogen in "learning" mode for MIDI events, I see the command going to my midi port, but Hydrogen isn't getting anything.
If I use PocketMID to send CC commands (like foot, or legato or any other CC command), then Hydrogen receives it.

I can even SEE the midi indicator flashing in Hydrogen when I send a play/stop signal...
So I KNOW that the MIDI signal is getting to Hydrogen... it's just not registering anything.

Any ideas?

**Comment by mauser** [comment]
Hi! 

ok, that is really strange.. Hydrogen  can log the midi messages if you're starting it from the commandline via `./hydrogen -VInfo
`. It would be interesting to see the output when you send the messages which are ignored by Hydrogen. Have you tried to send simple Note messages?

**Reply by rex007can** [reply]
That was a very good idea, however it doesn't work in Windows... Starting from command line with hydrogen.exe -V info doesn't do anything other than starting the program. No logs.

It does receive notes, I've tried that from Cakewalk and it seemed to be working. Just, no possible action on start/stop messages.

**Reply by rex007can** [reply]
Ha!
Wait!
On a hunch, I used Moba XTerm to launch it instead of a windows command prompt, low and behold, it actually logs in Moba...

So here are a few examples.
**These are if I send it a CC_001 (Modulation) messages:**
(Hydrogen logs)
```
(I) PortMidiDriver::handleMidiMessage [start of handleMidiMessage]
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] channel: 0
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] val1: 1
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] val2: 0
(I) PortMidiDriver::handleMidiMessage [start of handleMidiMessage]
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] channel: 0
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] val1: 1
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] val2: 0
(I) PortMidiDriver::handleMidiMessage [start of handleMidiMessage]
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] channel: 0
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] val1: 1
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] val2: 0
```

(Pocket Midi logs)
```
08:21:30:169 Port1 Ch 1 CC       B0 01 00 [001 Modulation]
08:21:30:332 Port1 Ch 1 CC       B0 01 00 [001 Modulation]
08:21:30:517 Port1 Ch 1 CC       B0 01 00 [001 Modulation]
08:21:30:679 Port1 Ch 1 CC       B0 01 00 [001 Modulation]
```

These are what's logged when I send a STOP message:
(Hydrogen logs)
```
(I) PortMidiDriver::handleMidiMessage [start of handleMidiMessage]
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] channel: 12
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] val1: 0
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] val2: 0
(E) PortMidiDriver::handleMidiMessage SYSTEM_EXCLUSIVE event not handled yet
(I) PortMidiDriver::handleMidiMessage [end of handleMidiMessage]
(I) PortMidiDriver::handleMidiMessage [start of handleMidiMessage]
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] channel: 12
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] val1: 0
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] val2: 0
(E) PortMidiDriver::handleMidiMessage SYSTEM_EXCLUSIVE event not handled yet
(I) PortMidiDriver::handleMidiMessage [end of handleMidiMessage]
(I) PortMidiDriver::handleMidiMessage [start of handleMidiMessage]
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] channel: 12
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] val1: 0
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] val2: 0
```

(Pocket midi)
```
08:26:47:959 Port1 ---- Stop     FC
08:26:48:129 Port1 ---- Stop     FC
08:26:48:925 Port1 ---- Stop     FC
08:26:49:133 Port1 ---- Stop     FC
```

**These are (cleaned up) logs from Hydrogen showing the commands sent from Cakewalk:**

```
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] channel: 11
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] val1: 0
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] val2: 0
(E) PortMidiDriver::handleMidiMessage SYSTEM_EXCLUSIVE event not handled yet
(I) PortMidiDriver::handleMidiMessage [end of handleMidiMessage]

(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] channel: 12
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] val1: 0
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] val2: 0
(E) PortMidiDriver::handleMidiMessage SYSTEM_EXCLUSIVE event not handled yet
(I) PortMidiDriver::handleMidiMessage [end of handleMidiMessage]

(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] channel: 0
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] val1: 0
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] val2: 64
(E) PortMidiDriver::handleMidiMessage PITCH_WHEEL event not handled yet
(I) PortMidiDriver::handleMidiMessage [end of handleMidiMessage]
```

**Comment by mauser** [comment]
Hi! 

the problem is that Hydrogen does not understand SYSTEM_EXCLUSIVE and PITCH_WEEL messages. Those message types are not implemented, also not on linux. Could you configure your software to send other command? Another option would be to use a midi mapping software, which maps the SYSTEM_EXCLUSIVE messages to CC, Note or Program Change messages.

**Reply by rex007can** [reply]
Well, when I send a standard STOP message
08:26:47:959 Port1 ---- Stop     FC

ALL that shows up in the Hydrogen logs is this and nothing else.

(I) PortMidiDriver::handleMidiMessage [start of handleMidiMessage]
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] channel: 12
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] val1: 0
(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] val2: 0
(E) PortMidiDriver::handleMidiMessage SYSTEM_EXCLUSIVE event not handled yet

I don't care about pitch-wheel, that's just dirt in the logs

(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] channel: 12 
This is stop

(I) PortMidiDriver::handleMidiMessage [handleMidiMessage] channel: 11
This is play
 etc.

**Comment by mauser** [comment]
Hi! 

Ok, i suppose i understand the problem now. Hydrogen indeed supports SYSEX messages, but only certain ones (like the STOP, PLAY MMC messages). But our portmidi driver (which we're using on windows) gets it wrong. I'm trying to fix that..

**Comment by nunosga** [comment]
Hi,
I'm trying to use a midi keyboard (Korg Microkey) to play / record pattenrs in hydrogen, but i can't get Hydrogen to work with the keyboard. (the korg works fine with studio one DAW).
Can you help me?
